lnrpc/routerrpc/router_backend.go

```diff
@@ -582,3 +623,49 @@ func ValidatePayReqExpiry(payReq *zpay32.Invoice) error {
 
 	return nil
 }
+
+// TlvRecordsToMap encodes a series of TLV records as raw key-value pairs in
+// the form of a map.
+func TlvRecordsToMap(records []tlv.Record) (map[uint64][]byte, error) {
+	tlvMap := make(map[uint64][]byte, len(records))
+
+	for _, record := range records {
+		var b bytes.Buffer
+		if err := record.Encode(&b); err != nil {
+			return nil, err
+		}
+
+		tlvMap[uint64(record.Type())] = b.Bytes()
+	}
+
+	return tlvMap, nil
+}
+
+// MapToTlvRecords encodes the passed TLV map as a series of regular tlv.Record
+// instances.
+func MapToTlvRecords(tlvMap map[uint64][]byte) ([]tlv.Record, error) {
+	// stubEncoder is a factory function that makes a stub tlv.Encoder out
+	// of a raw value. We can use this to make a record that can be encoded
+	// when we don't actually know it's true underlying value, and only it
+	// serialization.
+	stubEncoder := func(v []byte) tlv.Encoder {
+		return func(w io.Writer, val interface{}, buf *[8]byte) error {
+			_, err := w.Write(v)
+			return err
+		}
+	}
+
+	records := make([]tlv.Record, 0, len(tlvMap))
+	for k, v := range tlvMap {
+		// We don't pass in a decoder here since we don't actually know
+		// the type, and only expect this Record to be used for display
+		// and encoding purposes.
+		record := tlv.MakeStaticRecord(
+			tlv.Type(k), nil, uint64(len(v)), stubEncoder(v), nil,
```

Comment:
I think we should start enforcing the custom range for `k` here (>=2^16). To prevent unofficial pollution of the official types from the start.

Reply:
There isn't yet agreement on a custom range, nor where it starts (could just be the upper 4 bytes for example), so adding a constraint here would be premature.